server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -47,12 +71,20 @@ public StoredFieldsSpec storedFieldsSpec() {
 
             @Override
             public void process(HitContext hitContext) throws IOException {
-                Map<String, DocumentField> documentFields = fieldFetcher.fetch(hitContext.source(), hitContext.docId());
-                SearchHit hit = hitContext.hit();
-                for (Map.Entry<String, DocumentField> entry : documentFields.entrySet()) {
-                    hit.setDocumentField(entry.getKey(), entry.getValue());
-                }
+                final FieldFetcher.DocAndMetaFields fields = fieldFetcher.fetch(hitContext.source(), hitContext.docId());
+                hitContext.hit().addDocumentFields(fields.documentFields(), fields.metadataFields());
             }
         };
     }
+
+    /**
+     * _size must be excluded from metadata fields returned if the requests filters 'fields' using a wildcard.
+     * Put it another way, _size is returned only if requested explicitly through 'fields' or if the wildcard
+     * filter is used in combination with 'stored_fields'.
+     * @param fields List of fields requested as plain field names or including the wildcard character
+     * @return if we have to include the _size field or not
+     */
+    private static boolean includeSizeMetadataField(final List<FieldAndFormat> fields) {
+        return fields.stream().map(field -> field.field).noneMatch("*"::equals);
```

Comment:
I "extracted" this behavior from test `SizeMappingIT#testGetWithFields` which explicitly tests this behaviour testing `_size` requested explicitly or with a wildcard `*` for `fields` and `stored_fields`.

Reply:
I mentioned it above, I think this is problematic and we should have no mention of `_size` in here, as it is a custom metadata field that a plugin can provide. We need to be able to detect custom metadata fields in a generic manner, without knowing their names, otherwise they are not pluggable.